test/raft: test snapshot threshold

alecco opened the pull request:
Force snapshot with schema changes while server down. Then verify schema when bringing back up the server.

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/93d57acf074d2c775decf321076d86592a9c601f:

- simplify driver custom server connect as the other servers are down anyway

kbr-scylla commented on test/topology/test_snapshot.py:
```diff
@@ -0,0 +1,65 @@
+    manager.driver_close()
+    logger.debug("Starting C %s", server_c)
+    await manager.server_start(server_c.server_id)
+
```
I don't think there's a guarantee that the snapshot will reach C before A and B stop.
I see two options:
1. force a read barrier on C at this point, e.g. by doing a schema change - but it has to be done on C, so we would have to connect directly to C. You could use `wait_for_cql_and_get_hosts` from pylib.util which you can use to obtain a `Host` object for C, then pass it to `cql.run_async` as `host` parameter and execute a create table, for example.
2. Instead of doing schema changes with dead server then restarting it, **bootstrap** a new server. Bootstrap will definitely force a snapshot - the new server does not have any entries - and I think the test will also become a bit simpler.

alecco replied:
Went for option 2

alecco commented on raft/server.cc:
```diff
@@ -1160,6 +1160,12 @@ future<> server_impl::applier_fiber() {
+               utils::get_local_injector().inject("raft_server_snapshot_reduce_threshold",
+                   [this] { _config.snapshot_threshold = 3; _config.snapshot_trailing = 2; });
+               utils::get_local_injector().inject("raft_server_snapshot_restore_threshold",
+                   [this] { _config.snapshot_threshold = 1024; _config.snapshot_trailing = 200; });
```
Maybe we could do some kind of error injection feature where we change some value for a block of code and then change it back. Is there a way with C++ templates? WDYT?

kbr-scylla replied:
> Maybe we could do some kind of error injection feature where we change some value for a block of code and then change it back.

That's a pretty vague description - please explain in more detail. For example a piece of code presenting how this feature would be used

alecco replied:
I was thinking some member function of `class error_injection`, perhaps `.with<T>(lambda_init, lambda_actual_code)`. (in this case T is `struct server::configuration`).

If injection is enabled:
1. create a temporary T
2. run `lambda_init` on the original T
3. run `lambda_actual_code`
4. swaps back the `T`s

If injection is disabled:
1. run `lambda_init` on the original T

cc @kostja

alecco replied:
```cpp
#include <iostream>
#include <algorithm>
#include <functional>

using std::cout;

template <bool injection_enabled>
class error_injection {
    using handler_fun = std::function<void()>;
public:
    template <typename T>
    [[gnu::always_inline]]
    void with(T& v, handler_fun f_init, handler_fun f) {
        T tmp(v);            // tmp copy
        f_init();            // change orig
        f();                 // run with orig
        std::swap(v, tmp);   // revert
    }
};

template <>
class error_injection<false> {
    using handler_fun = std::function<void()>;
public:
    template <typename T>
    [[gnu::always_inline]]
    void with(T& v, handler_fun f_init, handler_fun f) {
        f();                 // run with orig
    }
};

int main() {
    auto err = error_injection<true>();
    int i = 1;
    err.with(i,
            [&i] { i = 9; },
            [&i] { cout << i << " - lambda\n"; }
    );
    cout << i << " - outside\n";

    return 1;
}
```

Seems to optimize out fine with clang (but not gcc?) https://godbolt.org/z/rrPs45aca

alecco replied:
@kbr-scylla  ping

kbr-scylla replied:
Ah, so you want to enable the injection temporarily only for the duration of a code block and then have it disabled automatically after the code block finishes. And the abstraction assumes that the error injection's only side effect is modifying a certain value, so it copies the value then restores it. Although I don't understand why `f_init` is called when the injection is disabled. We should not do anything if it's disabled. Preferably zero injection-related code should be executed.

Anyway, I don't think this fits into our use case. We don't want to revert the values automatically after a block of code. In your test we want the values to remain modified until the user sends another request to revert them (which you did with another injection). During this period the applier_fiber loop will execute many times.

I think we should simplify the problem and not do the revert at all. Let's change the values and that's it. We want them to remain changed for the whole duration of the test. Mark the cluster as dirty so it won't be reused by other test (coincidentally it's already marked as dirty by this test).

kostja replied:
@kbr-scylla shouldn't it be enough to change server configuration once at server start? This is exactly "one-off" use case.

kbr-scylla replied:
> @kbr-scylla shouldn't it be enough to change server configuration once at server start? This is exactly "one-off" use case.

You would have to somehow enable the injection before the server starts. There's no reliable way to do this

kostja replied:
We should add it, like, a command line flag to enable injection. Going forward we will need more of these start-time injections. WDYT?

kbr-scylla replied:
Makes sense

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4171/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/32e662500bfa7b0674234dbab73bf9f8b2a75c40:

- only set error injection once and mark cluster dirty (Kamil)
- fix test docstring
- use bootstrap server for snapshot transfer instead of taking server temporarily down
- change log level to `info`

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4179/

kbr-scylla commented on test/topology/test_snapshot.py:
```diff
@@ -0,0 +1,60 @@
+    errs = [inject_error_one_shot(manager.api, s.ip_addr, 'raft_server_snapshot_reduce_threshold')
+            for s in [server_a, server_b, server_c]]
+    await asyncio.gather(*errs)
+    # error injection is in Raft io_thread code path, wait for it to run
```
Now I see what you meant. You mean `applier_fiber` (the other one is `io_fiber`, not `io_thread`, but that's also not where you put the injection; you put it in `applier_fiber`).

Let's not do this like that. The sleep doesn't guarantee anything and it's just not an elegant way to solve the problem.

I suggest you do an additional schema change here. Once the schema change returns you're guaranteed that the fiber loop executed at least once on the node where you requested the schema change, because we add entry with `wait_type::applied`. We could actually move the injection to `io_fiber` - then it will be guaranteed even if we add_entry with `wait_type::committed`. Whatever you pick, this should be explained in a comment.

kbr-scylla commented on test/topology/test_snapshot.py:
```diff
@@ -0,0 +1,60 @@
+    await manager.driver_connect()
+    logger.info("Test DONE")
+
+    await manager.mark_dirty()
```
If the test fails right after enabling the injection, the cluster won't be dirty. Therefore this should be done before enabling the injection.

kbr-scylla commented on test/topology/test_snapshot.py:
```diff
@@ -0,0 +1,60 @@
+    await asyncio.gather(*[manager.server_stop_gracefully(s.server_id)
+                           for s in [server_a, server_b, server_c]])
+
+    logger.info("Driver connecting to C %s", server_c)
```
I think you wanted to verify schema on `D`. Also, write a comment what this dance with stopping servers and reconnecting driver is for. I had to guess that it's because you want to execute `verify_schema` on D. It's not obvious.

kbr-scylla commented on test/topology/test_snapshot.py:
```diff
@@ -0,0 +1,60 @@
+
+    await random_tables.verify_schema()
+
+    manager.driver_close()
```
Are these post-verify-schema steps actually necessary for something?

kbr-scylla replied:
Ah, right now maybe they are, you want to have a quorum of nodes alive to do the post-test check
(in another PR of mine I disable the post-test check for dirty clusters)

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/cf3b8d7edcd61b4fc3bcfe736312747dcbdc4c28:

- mark cluster dirty before changing snapshot threshold
- removed sleeps after error injection, schema changes trigger 2 snapshots 
- create table after threshold change
- reduce the schema changes (add column) to 3
- fix log string for server D doing schema verification 
- fixed test docstring and comment code

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4241/

kbr-scylla reviewed (approved):
Queued
